onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_api_helper.cc

```diff
@@ -0,0 +1,120 @@
+#include "nnapi_api_helper.h"
+
+#include "core/providers/nnapi/nnapi_builtin/builders/model_builder.h"
+#include "core/providers/nnapi/nnapi_builtin/nnapi_lib/nnapi_implementation.h"
+#include "core/common/logging/logging.h"
+
+#ifdef __ANDROID__
+#include <android/api-level.h>
+#endif
+
+namespace onnxruntime {
+namespace nnapi {
+
+int32_t GetNNAPIRuntimeFeatureLevel(const ::NnApi* nnapi) {
+  if (!nnapi)
+    return 0;
+  int32_t runtime_level = static_cast<int32_t>(nnapi->nnapi_runtime_feature_level);
+#ifdef __ANDROID__
+  int device_api_level = android_get_device_api_level();
+  runtime_level = (device_api_level < __ANDROID_API_S__) ? device_api_level : runtime_level;
+#endif
+  return runtime_level;
+}
+
+int32_t GetDeviceFeatureLevel(const ::NnApi* nnapi, const std::vector<ANeuralNetworksDevice*>& device_handles) {
+  if (!nnapi)
+    return 0;
+
+  int32_t target_feature_level = GetNNAPIRuntimeFeatureLevel(nnapi);
+
+  int64_t devices_feature_level = -1;
+
+  for (const auto* device_handle : device_handles) {
+    int64_t curr_device_feature_level = 0;
+    if (nnapi->ANeuralNetworksDevice_getFeatureLevel(device_handle, &curr_device_feature_level) != ANEURALNETWORKS_NO_ERROR) {
+      continue;
+    }
+
+    devices_feature_level = std::max(curr_device_feature_level, devices_feature_level);
```

Comment:
so this will get the max device feature level of all devices? if we want to use all devices, do we want to use the minimum one?

Reply:
From my understaning, we only want to run more ops on these devices, we don't care which device runs them.

So we can get the maximum level among these devices and get more ops suppported.

Reply:
NNAPI will try all devices which we passed into util it finds one.
In this example, NNAPI will check if this node is compatible with device A. If not, try device 2 then.

If we want to assign node to a specific device, say "device A", we should make devise A be our target device.

Reply:
Assumption: Everything other than the CPU device is faster than the NNAPI CPU reference implementation. Based on that, we want to use the highest level that does not result in fallback to CPU. 

Note that we're not assigning nodes to specific devices - NNAPI does that internally. So the level we use only says 'run with NNAPI', and not 'run with NNAPI using device X'.

Reply:
> Note that we're not assigning nodes to specific devices - NNAPI does that internally. So the level we use only says 'run with NNAPI', and not 'run with NNAPI using device X'.

right, I should update my question: If NNAPI then assigns a node to device A (which supports at most feature level 1) but the model requires feature level 2 (as determined by the NNAPI EP), that could be a problem, right?

Reply:
Is the "problem" implying runtime failure or in-efficient for extra data-copy between different devices?
We can't get a specific feature level for a specific Op. and we don't which device can run this node faster. So it's very hard for us to figure out a high-efficient strategy on how ops are allocated.

What we can do is to assign as more as possible nodes on NNAPI accelerators and let NNAPI do the execution planning.

Reply:
> Is the "problem" implying runtime failure or in-efficient for extra data-copy between different devices?

problem meaning runtime failure or incorrect results. I'm not sure if NNAPI will be smart about what device has what feature level support and only assign to a device the functionality it can handle. though I'm not so familiar with this so please do correct me if I am wrong.

couldn't find much about it in the docs, but they do mention this in the ANeuralNetworksDevice_getFeatureLevel() docs:
"Developers could decide whether or not the specified device should be used for a model that has certain feature requirements."

so basically, what I'm unsure about is whether it is ok to use a device with lower feature level support than what is expected by the model.

Reply:
>> problem meaning runtime failure or incorrect results. I'm not sure if NNAPI will be smart about what device has what feature level support and only assign to a device the functionality it can handle.

Yes, It can detect a device's ability and decide how to assign nodes.  It 
supplies an API named [ANeuralNetworksModel_getSupportedOperationsForDevices](https://developer.android.com/ndk/reference/group/neural-networks#aneuralnetworksmodel_getsupportedoperationsfordevices) for users to get status of which op is supported.